### DISCUSSION TITLE
inheritance

### DISCUSSION THREAD

**Discussion by stephanlevi** [prisma/prisma#2342] [Q&A]
## Problem
i need to use base modeling

## Solution
a data model which will inherit from another data model

## Alternatives
none i can think of

## Additional context
none

**Comment by kldzj** [comment]
Maybe you could further explain your use case, why can't you use relations that extend your base model?

**Comment by stephanlevi** [comment]
Hi Nikolai,

I could use a field called "base", which will be used as my base data_model
But it's less useable

Isn't there any other way to do it ?


*Sent with Shift
<https://tryshift.com/?utm_source=SentWithShift&utm_campaign=Sent%20with%20Shift%20Signature&utm_medium=Email%20Signature&utm_content=General%20Email%20Group>*

On Mon, May 4, 2020 at 2:03 AM Nikolai <notifications@github.com> wrote:

> Maybe you could further explain your use case, why can't you use relations
> that extend your base model?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prisma/prisma/issues/2342#issuecomment-623197692>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJREN3V4JXPBKENKUIGG3ZLRPXZ2LANCNFSM4MV4Z56A>
> .
>


-- 

Stephan Levi | VP R&D

+972-54-632-0037 | rbtree.io <http://www.rbtree.io/>

**Comment by kldzj** [comment]
I'm still having trouble to understand what (and why) you want to achieve. 

Are you trying to adapt an existing schema or are you able to design freely?

**Comment by stephanlevi** [comment]
ill give you an example
i would like to create two entities
person entity
student entity

i want the student entity to inherit from the person entity



*Sent with Shift
<https://tryshift.com/?utm_source=SentWithShift&utm_campaign=Sent%20with%20Shift%20Signature&utm_medium=Email%20Signature&utm_content=General%20Email%20Group>*

On Mon, May 4, 2020 at 12:20 PM Nikolai <notifications@github.com> wrote:

> I'm still having trouble to understand what (and why) you want to achieve.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prisma/prisma/issues/2342#issuecomment-623353971>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJREN3WK47I43UOL6IAEAHLRP2CHPANCNFSM4MV4Z56A>
> .
>


-- 

Stephan Levi | VP R&D

+972-54-632-0037 | rbtree.io <http://www.rbtree.io/>

**Comment by kldzj** [comment]
I'm not sure if there is a better way to achieve what you want, but your problem can be solved.

### Schema
```prisma
model Person {
  id        Int        @default(autoincrement()) @id
  title     String
  firstname String
  lastname  String
  birthdate DateTime
  teacher   Teacher?
  student   Student?
}

model Teacher {
  personId Int     @id
  person   Person  @relation(fields: [personId], references: [id])
  isProf   Boolean
}

model Student {
  personId Int    @id
  person   Person @relation(fields: [personId], references: [id])
  class    String
}
```

### JS
```javascript
await prisma.teacher.create({
  data: {
    isProf: true,
    person: {
      create: {
        title: 'test',
        firstname: 'test',
        lastname: 'test',
        birthdate: new Date().toISOString(),
      },
    },
  },
});

const persons = await prisma.person.findMany({ first: 1, include: { student: true, teacher: true } });
log({ persons });

const teacher = await prisma.teacher.findMany({ first: 1, include: { person: true } });
log({ teacher });

await prisma.disconnect();
```

### Result
```
{
    "persons": [
        {
            "id": 1,
            "title": "test",
            "firstname": "test",
            "lastname": "test",
            "birthdate": "2020-05-04T09:57:41.275Z",
            "student": null,
            "teacher": {
                "personId": 1,
                "isProf": true
            }
        }
    ]
}
```

```
{
    "teacher": [
        {
            "personId": 1,
            "isProf": true,
            "person": {
                "id": 1,
                "title": "test",
                "firstname": "test",
                "lastname": "test",
                "birthdate": "2020-05-04T09:57:41.275Z"
            }
        }
    ]
}
```

Edit: Removed the PersonType as it is probably not needed and just adds redundancy. The type can be determined by checking if relation fields (e.g. student/teacher) are available.

**Comment by stephanlevi** [comment]
first of all
i would like to thank you very much
you are the first serious response i got from prisma

regarding what you sent, this is more or less what i thought you said

can you explain this -

model Teacher {
  personId Int     @id
  person   Person  @relation(fields: [personId], references: [id])
  isProf   Boolean
}

the person field is a reference to the BASE - which is ok

why does the teacher needs the person_id field, and not just id ?

can you elaborate on the settings inside the RELATION plz


really appreciate it




*Sent with Shift
<https://tryshift.com/?utm_source=SentWithShift&utm_campaign=Sent%20with%20Shift%20Signature&utm_medium=Email%20Signature&utm_content=General%20Email%20Group>*

On Mon, May 4, 2020 at 1:03 PM Nikolai <notifications@github.com> wrote:

> I'm not sure if there is a better way to achieve what you want, but your
> problem can be solved.
> Schema
>
> model Person {  id        Int        @default(autoincrement()) @id  type      PersonType @default(student)  title     String  firstname String  lastname  String  birthdate DateTime  teacher   Teacher?  student   Student?}
> model Teacher {  personId Int     @id  person   Person  @relation(fields: [personId], references: [id])  isProf   Boolean}
> model Student {  personId Int    @id  person   Person @relation(fields: [personId], references: [id])  class    String}
> enum PersonType {  student  teacher  whatever}
>
> JS
>
> await prisma.teacher.create({
>   data: {
>     isProf: true,
>     person: {
>       create: {
>         title: 'test',
>         type: 'teacher',
>         firstname: 'test',
>         lastname: 'test',
>         birthdate: new Date().toISOString(),
>       },
>     },
>   },
> });
> const persons = await prisma.person.findMany({ first: 1, include: { student: true, teacher: true } });log({ persons });
> const teacher = await prisma.teacher.findMany({ first: 1, include: { person: true } });log({ teacher });
> await prisma.disconnect();
>
> Result
>
> {
>     "persons": [
>         {
>             "id": 1,
>             "type": "teacher",
>             "title": "test",
>             "firstname": "test",
>             "lastname": "test",
>             "birthdate": "2020-05-04T09:57:41.275Z",
>             "student": null,
>             "teacher": {
>                 "personId": 1,
>                 "isProf": true
>             }
>         }
>     ]
> }
>
> {
>     "teacher": [
>         {
>             "personId": 1,
>             "isProf": true,
>             "person": {
>                 "id": 1,
>                 "type": "teacher",
>                 "title": "test",
>                 "firstname": "test",
>                 "lastname": "test",
>                 "birthdate": "2020-05-04T09:57:41.275Z"
>             }
>         }
>     ]
> }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prisma/prisma/issues/2342#issuecomment-623373378>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJREN3UJH7KNJXYJB5HUAU3RP2HFNANCNFSM4MV4Z56A>
> .
>


-- 

Stephan Levi | VP R&D

+972-54-632-0037 | rbtree.io <http://www.rbtree.io/>

**Comment by kldzj** [comment]
Just to clarify, I'm not associated with prisma. You likely didn't receive a response yet because there are hundreds of other important issues that are very actively worked on by these lovely people.

Regarding the relations, as there should always at most be one Teacher or Student object per Person we can simply use the personId as primary key. Sure you could just create another ID for the related models, but the primary key will take care of the uniqueness for you so there never will be more than one Teacher or Student object per Person. To recap: this way you'll have the same effect as if you had an unrelated ID (primary key), as well as a foreign and a unique key on the personId reference.

Regarding the relation attribute, take a look at the [docs](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#the-relation-attribute).

Hope this helped.

**Comment by janpio** [comment]
(There also was a long weekend before the initial post and now)

**Comment by pantharshit00** [comment]
Related to https://github.com/prisma/prisma/issues/1291 and this previous rfc that we had for Prisma 1 : https://github.com/prisma/prisma1/issues/3407 and maybe https://github.com/prisma/prisma/issues/1644

Now I would like to know if you want way to reuse certain set of fields (think extending a type) or having interfaces that would make sure certain types follow same behavior(think in terms of interfaces and implementations).

In any case I don't think what you are asking are relations because I think you won't want to associate the types with each other but you want to have ability to copy fields between them. Correct me if that is not the case. 

It would be better if you can open a more detailed feature request with a detailed explanation of what exactly you want. An example schema design for your usecase will also help in understanding your exact use case.

**Comment by stephanlevi** [comment]
hi harshit,
i have three issues
1. i want to be able to write prisma code first
2. i want to be able to write data_models in JS or TS files
3. i want the ability to have data_model inheritance

*Sent with Shift
<https://tryshift.com/?utm_source=SentWithShift&utm_campaign=Sent%20with%20Shift%20Signature&utm_medium=Email%20Signature&utm_content=General%20Email%20Group>*

On Tue, May 5, 2020 at 2:50 PM Harshit Pant <notifications@github.com>
wrote:

> Related to #1291 <https://github.com/prisma/prisma/issues/1291> and this
> previous rfc that we had for Prisma 1 : prisma/prisma1#3407
> <https://github.com/prisma/prisma1/issues/3407> and maybe #1644
> <https://github.com/prisma/prisma/issues/1644>
>
> Now I would like to know if you want way to reuse certain set of fields
> (think extending a type) or having interfaces that would make sure certain
> types follow same behavior(think in terms of interfaces and
> implementations).
>
> In any case I don't think what you are asking are relations because I
> think you won't want to associate the types with each other but you want to
> have ability to copy fields between them. Correct me if that is not the
> case.
>
> It would be better if you can open a more detailed feature request with a
> detailed explanation of what exactly you want. An example schema design for
> your usecase will also help in understanding your exact use case.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prisma/prisma/issues/2342#issuecomment-624008276>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJREN3VOGGAFPI6SKE7R3FTRP74RPANCNFSM4MV4Z56A>
> .
>


-- 

Stephan Levi | VP R&D

+972-54-632-0037 | rbtree.io <http://www.rbtree.io/>

**Comment by pantharshit00** [comment]
> 1. i want to be able to write prisma code first
> 2. i want to be able to write data_models in JS or TS files

I think you opened https://github.com/prisma/prisma/issues/2343 for this request but that is request is a duplicate. Please always search through the issues of a repository before opening a new issue so that you can avoid opening duplicate issues.

> 3. i want the ability to have data_model inheritance

I will keep this issue for this feature request for now.

**Comment by stephanlevi** [comment]
sure, tnx

*Sent with Shift
<https://tryshift.com/?utm_source=SentWithShift&utm_campaign=Sent%20with%20Shift%20Signature&utm_medium=Email%20Signature&utm_content=General%20Email%20Group>*

On Tue, May 5, 2020 at 3:17 PM Harshit Pant <notifications@github.com>
wrote:

>
>    1. i want to be able to write prisma code first
>    2. i want to be able to write data_models in JS or TS files
>
> I think you opened #2343 <https://github.com/prisma/prisma/issues/2343>
> for this request but that is request is a duplicate. Please always search
> through the issues of a repository before opening a new issue so that you
> can avoid opening duplicate issues.
>
>
>    1. i want the ability to have data_model inheritance
>
> I will keep this issue for this feature request for now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prisma/prisma/issues/2342#issuecomment-624020461>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJREN3SRTQVS3HJY5YZPMKDRP77UFANCNFSM4MV4Z56A>
> .
>


-- 

Stephan Levi | VP R&D

+972-54-632-0037 | rbtree.io <http://www.rbtree.io/>
